test/metabase/query_processor/middleware/add_dimension_projections_test.clj

```diff
@@ -0,0 +1,181 @@
+                    [2 "Stout Burgers & Beers" 11 2 "Bar"]
+                    [3 "The Apple Pan" 11 2 "Bar"]
+                    [4 "Wurstküche" 29 2 "Baz"]
+                    [5 "Brite Spot Family Restaurant" 20 2 "Qux"]])
+      (update :columns conj "Foo")
+      (update :cols (fn [cols]
+                      (conj (mapv (fn [col] (dissoc col :dimensions :values)) cols)
+                            {:description nil,
+                             :id nil,
+                             :table_id nil,
+                             :expression-name "Foo",
+                             :source :fields,
+                             :name "Foo",
+                             :display_name "Foo",
+                             :target nil,
+                             :extra_info {}
+                             :remapped_from "CATEGORY_ID"
+                             :remapped_to nil})))
+      (update :cols #(mapv (fn [col]
+                             (if (= "CATEGORY_ID" (:name col))
+                               (assoc col
+                                 :remapped_to "Foo"
+                                 :remapped_from nil)
+                               col))
+                           %)))
+  (add-inline-remaps example-resultset))
+
+(def example-query
+  {:query
+   {:fields
+    [{:description "A unique internal identifier for the review. Should not be used externally.",
+      :base-type :type/BigInteger,
+      :table-id 4,
+      :special-type :type/PK,
+      :field-name "ID",
+      :dimensions [],
+      :field-display-name "ID",
+      :values [],
+      :visibility-type :normal,
+      :position 0,
+      :field-id 31,
+      :table-name "REVIEWS",
+      :schema-name "PUBLIC"}
+     {:description "The review the user left. Limited to 2000 characters.",
+      :base-type :type/Text,
+      :table-id 4,
+      :special-type :type/Description,
+      :field-name "BODY",
+      :dimensions [],
+      :field-display-name "Body",
+      :values [],
+      :visibility-type :normal,
+      :position 0,
+      :field-id 29,
+      :table-name "REVIEWS",
+      :schema-name "PUBLIC"}
+     {:field-id 32,
+      :field-name "PRODUCT_ID",
+      :field-display-name "Product ID",
+      :base-type :type/Integer,
+      :special-type :type/FK,
+      :visibility-type :normal,
+      :table-id 4,
+      :schema-name "PUBLIC",
+      :table-name "REVIEWS",
+      :position 0,
+      :fk-field-id nil,
+      :description "The product the review was for",
+      :parent-id nil,
+      :parent nil,
+      :remapped-from nil,
+      :remapped-to nil,
+      :dimensions {:id 2, :name "Product", :field_id 32, :human_readable_field_id 27, :type "external"},
+      :values []}]}})
+
+(expect
+  (-> example-query
+      ;; This needs to be fixed, we don't know the name of the new
+      ;; column until it gets resolved, so it doesn't get set here
+      #_(update-in [:query :fields 2] assoc :remapped-to "TITLE" :remapped-from nil)
```

Comment:
@salsakran @camsaul @tlrobinson 

The PR currently works for FK remapping by letting the foreign key get resolved like it would normally and that defines the name of the remapped field. Is that the right field name for this? For internal remapping we let the user decide the name of the field. Should the user also be able to decide the name of the FK remapped field?

If so, should that name be used just as the display_name of the column? Or as the name and display_name? Not sure if this plays into the answer or not, but remapped-to/remapped-from are two new fields that @tlrobinson and I added to indicate a source/destination field in a result set. Currently that field uses the name of the source/destination of the remapping.